gapii/cc/pack_encoder.cpp

```diff
@@ -295,11 +296,11 @@ void PackEncoderImpl::writeVarint(std::string& buffer, uint64_t value) {
 }
 
 uint64_t PackEncoderImpl::flushChunk(std::string& buffer, bool isTypeDefChunk) {
-  int64_t size = buffer.size();
-  std::string sizeBuffer;
-  writeZigzag(sizeBuffer, isTypeDefChunk ? -size : size);
-  mShared->writer->write(sizeBuffer);
-  mShared->writer->write(buffer);
+  const int64_t size = buffer.size();
```

Comment:
I don't think this is the right place for this. 
The PackEncoder should not be aware of the protocol, it is responsible "only" for taking objects being serialized, and dumping them to the stream. I would expect the stream itself to be the one responsible for the protocol.

Reply:
This part has nothing to do with the GAPII-GAPIS protocol. This was just an optimization, so that the pack_encoder would send each chunk with a single write command instead of two.

Reply:
aha, sorry, I misunderstood. I assumed it was with the header on the protocol. 
However, I am not sure if/how much this will actually be an optimization?
Now we have to allocate a string, then realloc a string as it expands. I would expect most of the time write() to be cheaper than an extra malloc/free cycle.

Reply:
In case of unbuffered transfer, each write will trigger a flush in the chunk_writer, so each part would be encapsulated with the GAPII protocol header and send out. If we want to avoid malloc/free cycles, we could use the same optimization here that I used in the chunk_writer.

Reply:
I won't hold this up on that account, but we probably want to consider optimizing for the common case (buffered) since unbuffered is the slow-path already.
That being said it might even be simpler to update the implementation of:
`bool ChunkWriterImpl::write(std::string& s) {`
to 
```c++
bool ChunkWriterImpl::write(std::intiializer_list<const std::string* s> strings) {
   if (mStreamGood) {
    for (auto* s: strings) {
       mBuffer.append(*s);
    }
    if (mNoBuffer || (mBuffer.size() >= kBufferSize)) {
      flush();
    }
  }

  return mStreamGood;
}
```

Reply:
Yep, that would be easier, but would break the StringWriter API and its permission to move the string(s).

Reply:
Yep non-const std::string* is fine. It never should have been std::string& in the first place. (Google style guide does not allow non-const refs as method parameters.